src/world/World.php

```diff
@@ -593,10 +604,120 @@ private function initRandomTickBlocksFromConfig(ServerConfigGroup $cfg) : void{
+	}
+
+	public function setWeather(WeatherType $type, int $duration = 6000) : void{
+		if($this->weather === $type){
+			$this->weatherDuration = $duration;
+			return;
+		}
+
+		$ev = new WeatherChangeEvent($this, $this->weather, $type);
+		$ev->call();
+		if($ev->isCancelled()){
+			return;
+		}
+
+		$this->weather = $type;
+		$this->weatherDuration = $duration;
+		$this->weatherTick = 0;
+
+		$this->applyWeather();
+		$this->broadcastWeatherPackets();
+	}
+
+	public function tickWeather() : void{
+		$this->weatherTick++;
+
+		if($this->weatherTick >= $this->weatherDuration && $this->weatherDuration > 0){
+			$this->autoChangeWeather();
+		}
+
+		if($this->weather === WeatherType::THUNDER && mt_rand(0, 200) === 0){
+			$players = $this->getPlayers();
+			if(count($players) > 0){
+				$target = $players[array_rand($players)];
+				$spawn = $target->getPosition();
+				$x = (int) $spawn->x + mt_rand(-30, 30);
+				$z = (int) $spawn->z + mt_rand(-30, 30);
+				$y = $this->getHighestBlockAt($x, $z) ?? max(64, (int) $spawn->y);
+
+				$this->spawnLightning(new Vector3($x, $y, $z));
+			}
+		}
+	}
+
+	private function autoChangeWeather() : void{
+		$roll = mt_rand(1, 100);
+		if($roll <= 70){
+			$this->setWeather(WeatherType::CLEAR, mt_rand(6000, 12000));
+		}elseif($roll <= 90){
+			$this->setWeather(WeatherType::RAIN, mt_rand(4000, 9000));
+		}else{
+			$this->setWeather(WeatherType::THUNDER, mt_rand(2000, 6000));
+		}
+	}
+
+	private function applyWeather() : void{
+		match($this->weather){
+    	WeatherType::CLEAR => (function(){
+	    	$this->setRainLevel(0.0);
+	    	$this->setThunderLevel(0.0);
+    	})(),
+    	WeatherType::RAIN => (function(){
+    		$this->setRainLevel(1.0);
+	    	$this->setThunderLevel(0.0);
+    	})(),
+    	WeatherType::THUNDER => (function(){
+    		$this->setRainLevel(1.0);
+	    	$this->setThunderLevel(1.0);
+    	})(),
+    };
+	}
+
+	private function broadcastWeatherPackets() : void{
+		$players = $this->getPlayers();
+		if(count($players) === 0){
+			return;
+		}
+
+		foreach($players as $player){
+			$pos = $player->getPosition();
+
+			[$rainEvent, $thunderEvent] = match($this->weather){
+				WeatherType::CLEAR => [
+					LevelEventPacket::create(LevelEvent::STOP_RAIN, 0, $pos),
+					LevelEventPacket::create(LevelEvent::STOP_THUNDER, 0, $pos),
+				],
+				WeatherType::RAIN => [
+					LevelEventPacket::create(LevelEvent::START_RAIN, 80000, $pos),
+					LevelEventPacket::create(LevelEvent::STOP_THUNDER, 0, $pos),
+				],
+				WeatherType::THUNDER => [
+					LevelEventPacket::create(LevelEvent::START_RAIN, 160000, $pos),
+					LevelEventPacket::create(LevelEvent::START_THUNDER, 200000, $pos),
+				],
+			};
+
+			NetworkBroadcastUtils::broadcastPackets([$player], [$rainEvent, $thunderEvent]);
+		}
+	}
+
+	public function spawnLightning(Vector3 $pos) : void{
+		$bolt = new LightningBolt(Location::fromObject($pos, $this));
+		$bolt->spawnToAll();
+		$this->lightnings[spl_object_id($bolt)] = $bolt;
+	}
```

Comment:
This shouldn't be needed, plugins that want to create lightning can do so as they would any other entity